src/hardware/link-mso19/api.c

```diff
@@ -53,124 +63,265 @@ static const uint64_t samplerates[] = {
 	SR_HZ(100),
 };
 
-static const char *trigger_slopes[2] = {
-	"r", "f",
+static const char *trigger_sources[] = {
+	"DSO", "LA", // "SPI", "I2C",
+};
+
+enum {
+	TRIGGER_SOURCE_DSO = 0,
+	TRIGGER_SOURCE_LA,
+	// TRIGGER_SOURCE_SPI,
+	// TRIGGER_SOURCE_I2C,
+};
+
+static const char *dso_trigger_slopes[] = {
+	"Rising", "Falling",
+};
+
+static const char *la_trigger_slopes[] = {
+	"F->T", "T->F",
+};
+
+enum {
+	TRIGGER_SLOPE_RISING = 0,
+	TRIGGER_SLOPE_FALLING,
+	TRIGGER_SLOPE_F_T = 0,
+	TRIGGER_SLOPE_T_F,
 };
 
+static const int32_t trigger_matches[] = {
+	SR_TRIGGER_ZERO,
+	SR_TRIGGER_ONE,
+};
+
+static const char *logic_thresholds[] = {
+	"1.2V Logic",
+	"1.5V Logic",
+	"1.8V Logic",
+	"2.5V Logic",
+	"3.0V Logic",
+	"3.3V/5.0V Logic",
+};
+
+/* Values taken from USB wireshark capture */
+static const uint16_t logic_threshold_values[] = {
+	0x600,
+	0x770,
+	0x8ff,
+	0xc70,
+	0xeff,
+	0xfff,
+};
+
+static void mso_update_trigger_slope(struct dev_context *devc)
+{
+	switch (devc->trigger_source) {
+	case TRIGGER_SOURCE_DSO:
+		switch (devc->dso_trigger_slope) {
+		case TRIGGER_SLOPE_RISING:
+			TRIG_UPDATE_EDGE(devc->ctltrig, TRIG_EDGE_RISING);
+			break;
+		case TRIGGER_SLOPE_FALLING:
+			TRIG_UPDATE_EDGE(devc->ctltrig, TRIG_EDGE_FALLING);
+			break;
+		}
+		break;
+	case TRIGGER_SOURCE_LA:
+		switch (devc->la_trigger_slope) {
+		case TRIGGER_SLOPE_F_T:
+			TRIG_UPDATE_EDGE(devc->ctltrig, TRIG_EDGE_F_T);
+			break;
+		case TRIGGER_SLOPE_T_F:
+			TRIG_UPDATE_EDGE(devc->ctltrig, TRIG_EDGE_T_F);
+			break;
+		}
+		break;
+	}
+}
+
+static void mso_limit_trigger_level(struct dev_context *devc)
+{
+	double max_level = 2.0 * devc->dso_probe_factor;
```

Comment:
As a convention in this project, all variables used in a function are declared at the start of that function first and used later.